js_modules/dagit/packages/core/src/instance/JobMenu.tsx

```diff
@@ -0,0 +1,104 @@
+import {gql, useLazyQuery} from '@apollo/client';
+import * as React from 'react';
+
+import {RunFragments} from '../runs/RunFragments';
+import {doneStatuses, failedStatuses} from '../runs/RunStatuses';
+import {useJobReExecution} from '../runs/useJobReExecution';
+import {ButtonWIP} from '../ui/Button';
+import {IconWIP} from '../ui/Icon';
+import {MenuWIP, MenuLink, MenuItemWIP} from '../ui/Menu';
+import {Popover} from '../ui/Popover';
+import {RepoAddress} from '../workspace/types';
+import {workspacePipelinePath} from '../workspace/workspacePath';
+
+import {OverviewJobFragment} from './types/OverviewJobFragment';
+import {RunReExecutionQuery} from './types/RunReExecutionQuery';
+
+interface Props {
+  job: OverviewJobFragment;
+  repoAddress: RepoAddress;
+}
+
+/**
+ * Lazily load more information about the last run for this job, then use that data to inform
+ * whether re-execution is possible.
+ */
+export const JobMenu = (props: Props) => {
+  const {job, repoAddress} = props;
+  const lastRun = job.runs[0];
+  const [fetchHasExecutionPlan, {data}] = useLazyQuery<RunReExecutionQuery>(RUN_RE_EXECUTION_QUERY);
+
+  const run = data?.pipelineRunOrError.__typename === 'Run' ? data?.pipelineRunOrError : null;
+
+  const fetchIfPossible = React.useCallback(() => {
+    if (lastRun.id) {
+      fetchHasExecutionPlan({variables: {runId: lastRun.id}});
+    }
+  }, [lastRun, fetchHasExecutionPlan]);
+
+  const canRunFromFailure = React.useMemo(
+    () => !!run && !!run.executionPlan && failedStatuses.has(run.status),
+    [run],
+  );
+
+  const onLaunch = useJobReExecution(run);
```

Comment:
Hmm, is there any way this could use more of `<RunActionButtons />`? I think this looks correct but I'm a little wary of having two different code paths that have to produce the same enabled/disabled state. Maybe we could use the code there to generate the list of `LaunchButtonConfigurations` and then tweak the title text? 

I think this looks correct but also needs to check for the `canLaunchPipelineReexecution` permission unless that's inferred from it's location.

Reply:
Good catch on permissions, I think we need to include that here.